--- mplapack_old/mplapack/reference/Cgesvj.cpp	2021-08-01 17:23:21.617787538 +0900
+++ mplapack/mplapack/reference/Cgesvj.cpp	2021-08-01 15:53:29.389284517 +0900
@@ -378,11 +378,11 @@
     sn = sqrt(sfmin / epsln);
     temp1 = sqrt(big / castREAL(n));
     if ((aapp <= sn) || (aaqq >= temp1) || ((sn <= aaqq) && (aapp <= temp1))) {
-        temp1 = min(big, temp1 / aapp);
+        temp1 = min(big, REAL(temp1 / aapp));
         //         AAQQ  = AAQQ*TEMP1
         //         AAPP  = AAPP*TEMP1
     } else if ((aaqq <= sn) && (aapp <= temp1)) {
-        temp1 = min(sn / aaqq, big / (aapp * sqrt(castREAL(n))));
+        temp1 = min(REAL(sn / aaqq), REAL(big / (aapp * sqrt(castREAL(n)))));
         //         AAQQ  = AAQQ*TEMP1
         //         AAPP  = AAPP*TEMP1
     } else if ((aaqq >= sn) && (aapp >= temp1)) {
@@ -390,7 +390,7 @@
         //         AAQQ  = AAQQ*TEMP1
         //         AAPP  = AAPP*TEMP1
     } else if ((aaqq <= sn) && (aapp >= temp1)) {
-        temp1 = min(sn / aaqq, big / (sqrt(castREAL(n)) * aapp));
+        temp1 = min(REAL(sn / aaqq), REAL(big / (sqrt(castREAL(n)) * aapp)));
         //         AAQQ  = AAQQ*TEMP1
         //         AAPP  = AAPP*TEMP1
     } else {
@@ -435,7 +435,7 @@
         nbl++;
     }
     //
-    blskip = pow2(kbl);
+    blskip = kbl * kbl;
     //[TP] BLKSKIP is a tuning parameter that depends on SWBAND and KBL.
     //
     rowskip = min((INTEGER)5, kbl);
@@ -600,7 +600,7 @@
                                 //
                                 //                           AAPQ = AAPQ * CONJG( CWORK(p) ) * CWORK(q)
                                 aapq1 = -abs(aapq);
-                                mxaapq = max(mxaapq, -aapq1);
+                                mxaapq = max(mxaapq, REAL(-aapq1));
                                 //
                                 //        TO rotate or NOT to rotate, THAT is the question ...
                                 //
@@ -632,9 +632,9 @@
                                                 Crot(mvl, &v[(p - 1) * ldv], 1, &v[(q - 1) * ldv], 1, cs, conj(ompq) * t);
                                             }
                                             //
-                                            sva[q - 1] = aaqq * sqrt(max(zero, one + t * apoaq * aapq1));
-                                            aapp = aapp * sqrt(max(zero, one - t * aqoap * aapq1));
-                                            mxsinj = max(mxsinj, abs(t));
+                                            sva[q - 1] = aaqq * sqrt(max(zero, REAL(one + t * apoaq * aapq1)));
+                                            aapp = aapp * sqrt(max(zero, REAL(one - t * aqoap * aapq1)));
+                                            mxsinj = max(mxsinj, REAL(abs(t)));
                                             //
                                         } else {
                                             //
@@ -645,9 +645,9 @@
                                             cs = sqrt(one / (one + t * t));
                                             sn = t * cs;
                                             //
-                                            mxsinj = max(mxsinj, abs(sn));
-                                            sva[q - 1] = aaqq * sqrt(max(zero, one + t * apoaq * aapq1));
-                                            aapp = aapp * sqrt(max(zero, one - t * aqoap * aapq1));
+                                            mxsinj = max(mxsinj, REAL(abs(sn)));
+                                            sva[q - 1] = aaqq * sqrt(max(zero, REAL(one + t * apoaq * aapq1)));
+                                            aapp = aapp * sqrt(max(zero, REAL(one - t * aqoap * aapq1)));
                                             //
                                             Crot(m, &a[(p - 1) * lda], 1, &a[(q - 1) * lda], 1, cs, conj(ompq) * sn);
                                             if (rsvec) {
@@ -663,7 +663,7 @@
                                         Clascl("G", 0, 0, aaqq, one, m, 1, &a[(q - 1) * lda], lda, ierr);
                                         Caxpy(m, -aapq, &cwork[(n + 1) - 1], 1, &a[(q - 1) * lda], 1);
                                         Clascl("G", 0, 0, one, aaqq, m, 1, &a[(q - 1) * lda], lda, ierr);
-                                        sva[q - 1] = aaqq * sqrt(max(zero, one - aapq1 * aapq1));
+                                        sva[q - 1] = aaqq * sqrt(max(zero, REAL(one - aapq1 * aapq1)));
                                         mxsinj = max(mxsinj, sfmin);
                                     }
                                     //           END IF ROTOK THEN ... ELSE
@@ -796,7 +796,7 @@
                                 //
                                 //                           AAPQ = AAPQ * CONJG(CWORK(p))*CWORK(q)
                                 aapq1 = -abs(aapq);
-                                mxaapq = max(mxaapq, -aapq1);
+                                mxaapq = max(mxaapq, REAL(-aapq1));
                                 //
                                 //        TO rotate or NOT to rotate, THAT is the question ...
                                 //
@@ -823,9 +823,9 @@
                                             if (rsvec) {
                                                 Crot(mvl, &v[(p - 1) * ldv], 1, &v[(q - 1) * ldv], 1, cs, conj(ompq) * t);
                                             }
-                                            sva[q - 1] = aaqq * sqrt(max(zero, one + t * apoaq * aapq1));
-                                            aapp = aapp * sqrt(max(zero, one - t * aqoap * aapq1));
-                                            mxsinj = max(mxsinj, abs(t));
+                                            sva[q - 1] = aaqq * sqrt(max(zero, REAL(one + t * apoaq * aapq1)));
+                                            aapp = aapp * sqrt(max(zero, REAL(one - t * aqoap * aapq1)));
+                                            mxsinj = max(mxsinj, REAL(abs(t)));
                                         } else {
                                             //
                                             //                 .. choose correct signum for THETA and rotate
@@ -837,9 +837,9 @@
                                             t = one / (theta + thsign * sqrt(one + theta * theta));
                                             cs = sqrt(one / (one + t * t));
                                             sn = t * cs;
-                                            mxsinj = max(mxsinj, abs(sn));
-                                            sva[q - 1] = aaqq * sqrt(max(zero, one + t * apoaq * aapq1));
-                                            aapp = aapp * sqrt(max(zero, one - t * aqoap * aapq1));
+                                            mxsinj = max(mxsinj, REAL(abs(sn)));
+                                            sva[q - 1] = aaqq * sqrt(max(zero, REAL(one + t * apoaq * aapq1)));
+                                            aapp = aapp * sqrt(max(zero, REAL(one - t * aqoap * aapq1)));
                                             //
                                             Crot(m, &a[(p - 1) * lda], 1, &a[(q - 1) * lda], 1, cs, conj(ompq) * sn);
                                             if (rsvec) {
@@ -856,7 +856,7 @@
                                             Clascl("G", 0, 0, aaqq, one, m, 1, &a[(q - 1) * lda], lda, ierr);
                                             Caxpy(m, -aapq, &cwork[(n + 1) - 1], 1, &a[(q - 1) * lda], 1);
                                             Clascl("G", 0, 0, one, aaqq, m, 1, &a[(q - 1) * lda], lda, ierr);
-                                            sva[q - 1] = aaqq * sqrt(max(zero, one - aapq1 * aapq1));
+                                            sva[q - 1] = aaqq * sqrt(max(zero, REAL(one - aapq1 * aapq1)));
                                             mxsinj = max(mxsinj, sfmin);
                                         } else {
                                             Ccopy(m, &a[(q - 1) * lda], 1, &cwork[(n + 1) - 1], 1);
@@ -864,7 +864,7 @@
                                             Clascl("G", 0, 0, aapp, one, m, 1, &a[(p - 1) * lda], lda, ierr);
                                             Caxpy(m, -conj(aapq), &cwork[(n + 1) - 1], 1, &a[(p - 1) * lda], 1);
                                             Clascl("G", 0, 0, one, aapp, m, 1, &a[(p - 1) * lda], lda, ierr);
-                                            sva[p - 1] = aapp * sqrt(max(zero, one - aapq1 * aapq1));
+                                            sva[p - 1] = aapp * sqrt(max(zero, REAL(one - aapq1 * aapq1)));
                                             mxsinj = max(mxsinj, sfmin);
                                         }
                                     }
