--- mplapack_old/mplapack/reference/Chbevd_2stage.cpp	2021-08-01 17:23:21.617787538 +0900
+++ mplapack/mplapack/reference/Chbevd_2stage.cpp	2021-08-01 15:53:29.389284517 +0900
@@ -76,8 +76,8 @@
         lhtrd = iMlaenv2stage(3, "Chetrd_hb2st", jobz, n, kd, ib, -1);
         lwtrd = iMlaenv2stage(4, "Chetrd_hb2st", jobz, n, kd, ib, -1);
         if (wantz) {
-            lwmin = 2 * pow2(n);
-            lrwmin = 1 + 5 * n + 2 * pow2(n);
+            lwmin = 2 * n * n;
+            lrwmin = 1 + 5 * n + 2 * n * n;
             liwmin = 3 + 5 * n;
         } else {
             lwmin = max(n, lhtrd + lwtrd);
