--- mplapack_old/mplapack/reference/Cheevd_2stage.cpp	2021-08-01 17:23:21.617787538 +0900
+++ mplapack/mplapack/reference/Cheevd_2stage.cpp	2021-08-01 15:53:29.389284517 +0900
@@ -91,7 +91,7 @@
             lwtrd = iMlaenv2stage(4, "Chetrd_2stage", jobz, n, kd, ib, -1);
             if (wantz) {
                 lwmin = 2 * n + n * n;
-                lrwmin = 1 + 5 * n + 2 * pow2(n);
+                lrwmin = 1 + 5 * n + 2 * n * n;
                 liwmin = 3 + 5 * n;
             } else {
                 lwmin = n + 1 + lhtrd + lwtrd;
