--- mplapack_old/mplapack/reference/Claein.cpp	2021-08-01 17:23:21.617787538 +0900
+++ mplapack/mplapack/reference/Claein.cpp	2021-08-01 15:53:29.393284571 +0900
@@ -87,7 +87,7 @@
     //
     rootn = sqrt(castREAL(n));
     growto = tenth / rootn;
-    nrmsml = max(one, eps3 * rootn) * smlnum;
+    nrmsml = max(one, REAL(eps3 * rootn)) * smlnum;
     //
     //     Form B = H - W*I (except that the subdiagonal elements are not
     //     stored).
