--- mplapack_old/mplapack/reference/Claed7.cpp	2021-08-01 17:23:21.617787538 +0900
+++ mplapack/mplapack/reference/Claed7.cpp	2021-08-01 15:53:29.393284571 +0900
@@ -128,7 +128,7 @@
     if (k != 0) {
         Rlaed9(k, 1, k, n, d, &rwork[iq - 1], k, rho, &rwork[idlmda - 1], &rwork[iw - 1], &qstore[qptr[curr - 1] - 1], k, info);
         Clacrm(qsiz, k, work, qsiz, &qstore[qptr[curr - 1] - 1], k, q, ldq, &rwork[iq - 1]);
-        qptr[(curr + 1) - 1] = qptr[curr - 1] + pow2(k);
+        qptr[(curr + 1) - 1] = qptr[curr - 1] + k * k;
         if (info != 0) {
             return;
         }
