--- mplapack_old/mplapack/reference/Claqps.cpp	2021-08-01 17:23:21.617787538 +0900
+++ mplapack/mplapack/reference/Claqps.cpp	2021-08-01 15:53:29.397284626 +0900
@@ -159,7 +159,7 @@
                     //                 Lapack Working Note 176.
                     //
                     temp = abs(a[(rk - 1) + (j - 1) * lda]) / vn1[j - 1];
-                    temp = max(zero, (one + temp) * (one - temp));
+                    temp = max(zero, REAL((one + temp) * (one - temp)));
                     temp2 = temp * pow2((vn1[j - 1] / vn2[j - 1]));
                     if (temp2 <= tol3z) {
                         vn2[j - 1] = castREAL(lsticc);
