--- mplapack_old/mplapack/reference/Cpstrf.cpp	2021-08-01 17:23:21.621787593 +0900
+++ mplapack/mplapack/reference/Cpstrf.cpp	2021-08-01 15:53:29.401284681 +0900
@@ -144,7 +144,7 @@
                 //              holds dot products
                 //
                 for (i = k; i <= n; i = i + 1) {
-                    work[i - 1] = 0;
+                    work[i - 1] = 0.0;
                 }
                 //
                 for (j = k; j <= k + jb - 1; j = j + 1) {
@@ -234,7 +234,7 @@
                 //              holds dot products
                 //
                 for (i = k; i <= n; i = i + 1) {
-                    work[i - 1] = 0;
+                    work[i - 1] = 0.0;
                 }
                 //
                 for (j = k; j <= k + jb - 1; j = j + 1) {
