--- mplapack_old/mplapack/reference/Csyequb.cpp	2021-08-01 17:23:21.621787593 +0900
+++ mplapack/mplapack/reference/Csyequb.cpp	2021-08-01 15:53:29.401284681 +0900
@@ -240,7 +240,7 @@
     base = Rlamch("B");
     u = one / log(base);
     for (i = 1; i <= n; i = i + 1) {
-        s[i - 1] = pow(base, castREAL(u * log(s[i - 1] * t)));
+        s[i - 1] = pow(base, REAL(u * log(s[i - 1] * t)));
         smin = min(smin, s[i - 1]);
         smax = max(smax, s[i - 1]);
     }
