--- mplapack_old/mplapack/reference/Ctrevc.cpp	2021-08-01 17:23:21.621787593 +0900
+++ mplapack/mplapack/reference/Ctrevc.cpp	2021-08-01 15:53:29.405284737 +0900
@@ -168,7 +168,7 @@
                     goto statement_80;
                 }
             }
-            smin = max(ulp * (abs1(t[(ki - 1) + (ki - 1) * ldt])), smlnum);
+            smin = max(REAL(ulp * (abs1(t[(ki - 1) + (ki - 1) * ldt]))), smlnum);
             //
             work[1 - 1] = cmone;
             //
@@ -238,7 +238,7 @@
                     goto statement_130;
                 }
             }
-            smin = max(ulp * (abs1(t[(ki - 1) + (ki - 1) * ldt])), smlnum);
+            smin = max(REAL(ulp * (abs1(t[(ki - 1) + (ki - 1) * ldt]))), smlnum);
             //
             work[n - 1] = cmone;
             //
