--- mplapack_old/mplapack/reference/iMlaenv.cpp	2021-08-01 17:23:21.625787646 +0900
+++ mplapack/mplapack/reference/iMlaenv.cpp	2021-08-01 15:53:29.429285060 +0900
@@ -30,7 +30,7 @@
 #include <mplapack.h>
 #include <string.h>
 
-#define subnamlen 16
+#define subnamlen 17
 
 INTEGER iMlaenv(INTEGER const ispec, const char *name, const char *opts, INTEGER const n1, INTEGER const n2, INTEGER const n3, INTEGER const n4) {
     INTEGER return_value = 0;
@@ -50,24 +50,6 @@
     INTEGER nx = 0;
     INTEGER name_len;
     //
-    //  -- LAPACK auxiliary routine --
-    //  -- LAPACK is a software package provided by Univ. of Tennessee,    --
-    //  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--
-    //
-    //     .. Scalar Arguments ..
-    //     ..
-    //
-    //  =====================================================================
-    //
-    //     .. Local Scalars ..
-    //     ..
-    //     .. Intrinsic Functions ..
-    //     ..
-    //     .. External Functions ..
-    //     ..
-    //     .. Executable Statements ..
-    //
-
     switch (ispec) {
     case 1:
         goto L10;
@@ -113,8 +95,9 @@
     //     Convert NAME to upper case if the first character is lower case.
     //
     return_value = 1;
-    name_len = min((int)strlen(name), subnamlen);
-    strncpy(subnam, name, name_len);
+    //
+    // name_len = min((INTEGER)strlen(name), (INTEGER)subnamlen);
+    strcpy(subnam, name);
     ic = *subnam;
 
     for (int i = 0; i < strlen(subnam); i++) {
