--- mplapack_old/mplapack/reference/Rlarrd.cpp	2021-08-01 17:23:21.621787593 +0900
+++ mplapack/mplapack/reference/Rlarrd.cpp	2021-08-01 15:53:29.417284898 +0900
@@ -193,7 +193,7 @@
         //        RANGE='I': Compute an interval containing eigenvalues
         //        IL through IU. The initial interval [GL,GU] from the global
         //        Gerschgorin bounds GL and GU is refined by Rlaebz.
-        itmax = castREAL((log(tnorm + pivmin) - log(pivmin)) / log(two)) + 2;
+        itmax = castINTEGER((log(tnorm + pivmin) - log(pivmin)) / log(two)) + 2;
         if (itmax >= 1024)
             itmax = 1024; // XXX itmax can be too large for MPFR (=10^8)
         work[(n + 1) - 1] = gl;
