--- mplapack_old/mplapack/reference/Rlarrv.cpp	2021-08-01 17:23:21.621787593 +0900
+++ mplapack/mplapack/reference/Rlarrv.cpp	2021-08-01 15:53:29.417284898 +0900
@@ -390,10 +390,10 @@
                     //                 However, we only allow the gaps to become greater since
                     //                 this is what should happen when we decrease WERR
                     if (oldfst > 1) {
-                        wgap[(wbegin + oldfst - 2) - 1] = max(wgap[(wbegin + oldfst - 2) - 1], w[(wbegin + oldfst - 1) - 1] - werr[(wbegin + oldfst - 1) - 1] - w[(wbegin + oldfst - 2) - 1] - werr[(wbegin + oldfst - 2) - 1]);
+                        wgap[(wbegin + oldfst - 2) - 1] = max(wgap[(wbegin + oldfst - 2) - 1], REAL(w[(wbegin + oldfst - 1) - 1] - werr[(wbegin + oldfst - 1) - 1] - w[(wbegin + oldfst - 2) - 1] - werr[(wbegin + oldfst - 2) - 1]));
                     }
                     if (wbegin + oldlst - 1 < wend) {
-                        wgap[(wbegin + oldlst - 1) - 1] = max(wgap[(wbegin + oldlst - 1) - 1], w[(wbegin + oldlst) - 1] - werr[(wbegin + oldlst) - 1] - w[(wbegin + oldlst - 1) - 1] - werr[(wbegin + oldlst - 1) - 1]);
+                        wgap[(wbegin + oldlst - 1) - 1] = max(wgap[(wbegin + oldlst - 1) - 1], REAL(w[(wbegin + oldlst) - 1] - werr[(wbegin + oldlst) - 1] - w[(wbegin + oldlst - 1) - 1] - werr[(wbegin + oldlst - 1) - 1]));
                     }
                     //                 Each time the eigenvalues in WORK get refined, we store
                     //                 the newly found approximation with all shifts applied in W
@@ -455,7 +455,7 @@
                         //                    in W might be of the same order so that gaps are not
                         //                    exhibited correctly for very close eigenvalues.
                         if (newfst == 1) {
-                            lgap = max(zero, w[wbegin - 1] - werr[wbegin - 1] - vl);
+                            lgap = max(zero, REAL(w[wbegin - 1] - werr[wbegin - 1] - vl));
                         } else {
                             lgap = wgap[(wbegin + newfst - 2) - 1];
                         }
@@ -742,10 +742,10 @@
                         //                    to WERR being too crude.)
                         if (!eskip) {
                             if (k > 1) {
-                                wgap[windmn - 1] = max(wgap[windmn - 1], w[windex - 1] - werr[windex - 1] - w[windmn - 1] - werr[windmn - 1]);
+                                wgap[windmn - 1] = max(wgap[windmn - 1], REAL(w[windex - 1] - werr[windex - 1] - w[windmn - 1] - werr[windmn - 1]));
                             }
                             if (windex < wend) {
-                                wgap[windex - 1] = max(savgap, w[windpl - 1] - werr[windpl - 1] - w[windex - 1] - werr[windex - 1]);
+                                wgap[windex - 1] = max(savgap, REAL(w[windpl - 1] - werr[windpl - 1] - w[windex - 1] - werr[windex - 1]));
                             }
                         }
                         idone++;
