--- mplapack_old/mplapack/reference/Rlarrb.cpp	2021-08-01 17:23:21.621787593 +0900
+++ mplapack/mplapack/reference/Rlarrb.cpp	2021-08-01 15:53:29.417284898 +0900
@@ -230,7 +230,7 @@
     for (i = ifirst + 1; i <= ilast; i = i + 1) {
         k = 2 * i;
         ii = i - offset;
-        wgap[(ii - 1) - 1] = max(zero, w[ii - 1] - werr[ii - 1] - w[(ii - 1) - 1] - werr[(ii - 1) - 1]);
+        wgap[(ii - 1) - 1] = max(zero, REAL(w[ii - 1] - werr[ii - 1] - w[(ii - 1) - 1] - werr[(ii - 1) - 1]));
     }
     //
     //     End of Rlarrb
