--- mplapack_old/mplapack/reference/Rpstrf.cpp	2021-08-01 17:23:21.621787593 +0900
+++ mplapack/mplapack/reference/Rpstrf.cpp	2021-08-01 15:53:29.421284953 +0900
@@ -145,7 +145,7 @@
                 //              holds dot products
                 //
                 for (i = k; i <= n; i = i + 1) {
-                    work[i - 1] = 0;
+                    work[i - 1] = 0.0;
                 }
                 //
                 for (j = k; j <= k + jb - 1; j = j + 1) {
@@ -228,7 +228,7 @@
                 //              holds dot products
                 //
                 for (i = k; i <= n; i = i + 1) {
-                    work[i - 1] = 0;
+                    work[i - 1] = 0.0;
                 }
                 //
                 for (j = k; j <= k + jb - 1; j = j + 1) {
