--- mplapack_old/mplapack/reference/Rtgsna.cpp	2021-08-01 17:23:21.621787593 +0900
+++ mplapack/mplapack/reference/Rtgsna.cpp	2021-08-01 15:53:29.425285008 +0900
@@ -320,7 +320,7 @@
                     Rtgsyl("N", difdri, n2, n1, &work[(n * n1 + n1 + 1) - 1], n, work, n, &work[(n1 + 1) - 1], n, &work[(n * n1 + n1 + i) - 1], n, &work[i - 1], n, &work[(n1 + i) - 1], n, scale, dif[ks - 1], &work[(iz + 1) - 1], lwork - 2 * n * n, iwork, ierr);
                     //
                     if (pair) {
-                        dif[ks - 1] = min({max(one, alprqt) * dif[ks - 1], cond});
+                        dif[ks - 1] = min(REAL(max(one, alprqt) * dif[ks - 1]), cond);
                     }
                 }
             }
